rust/lance-file/src/reader.rs

```diff
@@ -0,0 +1,1800 @@
+    deletion_vector: Option<Arc<DeletionVector>>,
+
+    /// Page table for statistics
+    stats_page_table: Arc<Option<PageTable>>,
+}
+
+impl std::fmt::Debug for FileReader {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "FileReader(fragment={}, path={})",
+            self.fragment_id,
+            self.object_reader.path()
+        )
+    }
+}
+
+impl FileReader {
+    /// Open file reader
+    ///
+    /// Open the file at the given path using the provided object store.
+    ///
+    /// The passed fragment ID determines the first 32-bits of the row IDs.
+    ///
+    /// If a manifest is passed in, it will be used to load the schema and dictionary.
+    /// This is typically done if the file is part of a dataset fragment. If no manifest
+    /// is passed in, then it is read from the file itself.
+    ///
+    /// The session passed in is used to cache metadata about the file. If no session
+    /// is passed in, there will be no caching.
+    #[instrument(level = "debug", skip(object_store, manifest, session))]
+    pub async fn try_new_with_fragment(
+        object_store: &ObjectStore,
+        path: &Path,
+        fragment_id: u64,
+        manifest: Option<&Manifest>,
+        session: Option<&FileMetadataCache>,
+    ) -> Result<Self> {
+        let object_reader = object_store.open(path).await?;
+        let is_dataset = manifest.is_some();
+
+        let metadata = Self::read_metadata(object_reader.as_ref(), session).await?;
+
+        // m is either
+        // a ref if passed in as Some(&manifest), or
+        // a value we read in the else block
+        // since we don't want to clone the manifest and force everything into a value
+        // we need this m variable here so the else block has a place to store the read
+        // manifest value. If we declare this var in the else block, it won't live long
+        // enough and the returned ref will be invalid.
+        let mut m: Manifest;
+        let manifest = if let Some(m) = manifest {
+            m
+        } else {
+            m = read_struct(object_reader.as_ref(), metadata.manifest_position.unwrap()).await?;
+            m.schema.load_dictionary(object_reader.as_ref()).await?;
+            &m
+        };
+
+        // read the fragment metadata so we can handle deletion
+        let fragment = if is_dataset {
+            let fragment = manifest
+                .fragments
+                .iter()
+                .find(|frag| frag.id == fragment_id)
+                .map(|f| f.to_owned())
+                .ok_or(Error::IO {
+                    message: format!("Fragment {} not found in manifest", fragment_id),
+                    location: location!(),
+                })?;
+            Some(fragment)
+        } else {
+            None
+        };
+
+        let page_table = async {
+            Self::load_from_cache(session, path, |_| async {
```

Comment:
Did you intend to replace the `load_from_cache` calls with `get_or_insert` which you implemented in this PR?

Reply:
That was my original goal, for DRY, since the `FragmentReader` needs to call the same path.

However, in `FragmentReader` we have `&FileMetadataCache` and here in `FileReader` we have `Option<&FileMetadataCache>`.  So `Self::load_from_cache` should be calling `get_or_insert` if the cache is present.